src/plugins/schema.js

```diff
@@ -0,0 +1,304 @@
+import Schema from '../models/schema'
+import Text from '../models/text'
+import { List } from 'immutable'
+
+/*
+    This module contains the default schema to normalize documents
+ */
+
+function isInlineVoid(node) {
+  return (node.kind == 'inline' && node.isVoid)
+}
+
+/**
+ * A default schema rule to only allow block nodes in documents.
+ *
+ * @type {Object}
+ */
+
+const DOCUMENT_CHILDREN_RULE = {
+  match: (node) => {
+    return node.kind == 'document'
+  },
+  validate: (document) => {
+    const { nodes } = document
+    const invalids = nodes.filter(n => n.kind != 'block')
+    return invalids.size ? invalids : null
+  },
+  normalize: (transform, document, invalids) => {
+    return invalids.reduce((t, n) => t.removeNodeByKey(n.key, { normalize: false }), transform)
+  }
+}
+
+/**
+ * A default schema rule to only allow block, inline and text nodes in blocks.
+ *
+ * @type {Object}
+ */
+
+const BLOCK_CHILDREN_RULE = {
+  match: (node) => {
+    return node.kind == 'block'
+  },
+  validate: (block) => {
+    const { nodes } = block
+    const invalids = nodes.filter(n => n.kind != 'block' && n.kind != 'inline' && n.kind != 'text')
+    return invalids.size ? invalids : null
+  },
+  normalize: (transform, block, invalids) => {
+    return invalids.reduce((t, n) => t.removeNodeByKey(n.key, { normalize: false }), transform)
+  }
+}
+
+/**
+ * A default schema rule to have at least one text node in blocks/inlines
+ *
+ * @type {Object}
+ */
+
+const MIN_TEXT_RULE = {
+    match: (object) => {
+      return object.kind == 'block' || object.kind == 'inline'
+    },
+    validate: (node) => {
+      const { nodes } = node
+      return nodes.size === 0 ? true : null
+    },
+    normalize: (transform, node) => {
+      return transform.insertNodeByKey(node.key, 0, Text.create(), { normalize: false })
+    }
+}
+
+/**
+ * A default schema rule to only allow inline and text nodes in inlines.
+ *
+ * @type {Object}
+ */
+
+const INLINE_CHILDREN_RULE = {
+  match: (object) => {
+    return object.kind == 'inline'
+  },
+  validate: (inline) => {
+    const { nodes } = inline
+    const invalids = nodes.filter(n => n.kind != 'inline' && n.kind != 'text')
+    return invalids.size ? invalids : null
+},
+  normalize: (transform, inline, invalids) => {
+    return invalids.reduce((t, n) => t.removeNodeByKey(n.key, { normalize: false }), transform)
+  }
+}
+
+/**
+ * A default schema rule to ensure that inline nodes are not empty
```

Comment:
I added a comment, which is true for any normalization rule that remove nodes.
> This rule is applied to all blocks, because when they contain an
empty inline, we need to remove the inline from that parent
block. If `validate` was to be memoized, it should be against the
parent node, not the inline themselves.